### PR TITLE
Client cert renegotiation in http

### PR DESCRIPTION
**Pull request by fAuernigg**
Add server support for certificate renegotiation.
Fix tls state machine on client and server for tls1.2 renegotiation.

Feature Commits:
tls: handle client hello/finished state during tls 1.2 renegotiation
tls: handle server done state during tls 1.2 renegotiation

Commits with new Apis:
tls, http: add tls and http apis for renegotiation
        Add tls_set_conn_verify_client_handler,
        tls_renegotiate, tls_disable_session_on_reneg,
        tls_set_posthandshake_auth, tls_conn_verify_client_post_handshake,
        tls_conn_version
        Add http_sock_tls

Commit to fix retest on win32:
http: fix read_file on win32 (incorrect get fsize)

### CODE REVIEW

File: include/re_tls.h [hunk 1]
```diff
@@ -74,13 +88,23 @@ bool tls_get_session_reuse(const struct tls_conn *tc);
 int tls_reuse_session(const struct tls_conn *tc);
 bool tls_session_reused(const struct tls_conn *tc);
 int tls_update_sessions(const struct tls_conn *tc);
+#if !defined(LIBRESSL_VERSION_NUMBER)
+int tls_renegotiate(const struct tls_conn *tc);
+int tls_disable_session_on_reneg(struct tls_conn *tc);
+int tls_set_posthandshake_auth(struct tls *tls, int enabled);
+#endif
```
[the remark this answers is not in the document]
Comment (sreimers) [reply include/re_tls.h:95]: BTW: any chance to make these new apis LibreSSL compatible?

File: include/re_tls.h [hunk 2]
```diff
@@ -8,6 +8,10 @@ struct tls;
 struct tls_conn;
 struct tcp_conn;
 struct udp_sock;
+#if !defined(LIBRESSL_VERSION_NUMBER)
+struct x509_store_ctx_st;
+typedef struct x509_store_ctx_st X509_STORE_CTX;
```
Comment (sreimers) [inline include/re_tls.h:13]: The TLS api should be generic to not to be bound to a specific ssl libary.

File: include/re_tls.h [hunk 3]
```diff
@@ -28,6 +28,10 @@ enum tls_keytype {
 	TLS_KEYTYPE_EC,
 };
 
+struct tls_conn_d {
+	int (*verify_cb) (int ok, void *d);
+	void *verify_d;
```
Comment (cspiel1) [inline include/re_tls.h:33]: rename to `void *arg`

File: include/re_tls.h [hunk 4]
```diff
@@ -46,6 +50,10 @@ int tls_set_certificate_der(struct tls *tls, enum tls_keytype keytype,
 			    const uint8_t *key, size_t len_key);
 int tls_set_certificate(struct tls *tls, const char *cert, size_t len);
 void tls_set_verify_client(struct tls *tls);
+void tls_set_verify_client_trust_all(struct tls *tls);
+int tls_set_conn_verify_client_handler(struct tls_conn *tc, int depth,
+	int (*cb) (int ok, void *d), void *d);
```
Comment (cspiel1) [inline include/re_tls.h:55]: rename to
```c
int tls_set_verify_client_handler(struct tls_conn *tc, int depth,
	int (*verifyh) (int ok, void *arg), void *arg);
```

File: include/re_tls.h [hunk 5]
```diff
@@ -74,13 +82,19 @@ bool tls_get_session_reuse(const struct tls_conn *tc);
...
+int tls_disable_session_on_reneg(struct tls_conn *tc);
+int tls_set_posthandshake_auth(struct tls *tls, int enabled);
 
 /* TCP */
 
 int tls_conn_change_cert(struct tls_conn *tc, const char *file);
 int tls_start_tcp(struct tls_conn **ptc, struct tls *tls,
 		  struct tcp_conn *tcp, int layer);
 
+int tls_conn_verify_client_post_handshake(struct tls_conn *tc);
```
Comment (cspiel1) [inline include/re_tls.h:95]: rename
```c
int tls_verify_client_post_handshake(struct tls_conn *tc);
```

File: include/re_tls.h [hunk 6]
```diff
@@ -74,13 +82,19 @@ bool tls_get_session_reuse(const struct tls_conn *tc);
...
+const char* tls_conn_version(struct tls_conn *tc);
```
Comment (cspiel1) [inline include/re_tls.h:96]: rename to
```c
const char* tls_version(struct tls_conn *tc);
```

File: src/http/client.c [hunk 7]
```diff
@@ -741,28 +741,28 @@ static int read_file(char **pbuf, const char *path)
...
 		fclose(f);
 		return ENOMEM;
 	}
 
-	n = fread(buf, 1, s, f);
-	fclose(f);
-	if (n < (size_t)s) {
-		mem_deref(buf);
-		return EIO;
+	while ((n=fread(buf+idx, 1, PEMBUF_SIZE, f)) == PEMBUF_SIZE) {
```
Comment (cspiel1) [inline src/http/client.c:751]: maybe we should use the portable function `fs_fopen()`

File: src/http/client.c [hunk 8]
```diff
@@ -741,28 +741,28 @@ static int read_file(char **pbuf, const char *path)
...
+		idx += n;
+		nbuf = mem_realloc(buf, idx + PEMBUF_SIZE);
+		if (!nbuf) {
+			DEBUG_WARNING("Out of memory while reading cert file"\
+				"'%s',\n", path);
+			mem_deref(buf);
+			return ENOMEM;
+		}
+		buf = nbuf;
+		memset(buf+idx, 0, PEMBUF_SIZE);
```
Comment (cspiel1) [inline src/http/client.c:761]: We should use mbuf:
```c
	struct mbuf *cert;
	struct mbuf *key;
```
Please try this patch:

File: src/tls/openssl/tls.c [hunk 9]
```diff
@@ -39,6 +40,7 @@ struct session_reuse {
 	struct hash *ht_sessions;
 };
 
+
```
Comment (cspiel1) [inline src/tls/openssl/tls.c:43]: unrelated

File: src/tls/openssl/tls.c [hunk 10]
```diff
@@ -61,6 +63,7 @@ struct tls_cert {
 	char *host;
 };
 
+
```
Comment (cspiel1) [inline src/tls/openssl/tls.c:66]: unrelated

File: src/tls/openssl/tls.c [hunk 11]
```diff
@@ -936,6 +956,120 @@ void tls_set_verify_client(struct tls *tls)
 }
 
 
+/**
+ * Set TLS server context to request certificate from peer
+ * and set trust all certificates of peer.
+ *
+ * @deprecated Use tls_set_verify_peer_trust_all instead
```
Comment (cspiel1) [inline src/tls/openssl/tls.c:963]: please prepare also a PR for baresip to replace the deprecated function:
```
./modules/dtls_srtp/dtls_srtp.c:534:	tls_set_verify_client(tls);
```

File: src/tls/openssl/tls_tcp.c [hunk 12]
```diff
@@ -28,6 +28,7 @@
 struct tls_conn {
 	SSL *ssl;             /* inheritance */
 	struct tls *tls;      /* inheritance */
+	struct tls_conn_d cd; /* inheritance */
```
Comment (cspiel1) [inline src/tls/openssl/tls_tcp.c:31]: inheritance?

File: src/tls/openssl/tls_udp.c [hunk 13]
```diff
@@ -60,6 +60,7 @@ struct dtls_sock {
 struct tls_conn {
 	SSL *ssl;             /* inheritance */
 	struct tls *tls;      /* inheritance */
+	struct tls_conn_d cd; /* inheritance */
```
Comment (cspiel1) [inline src/tls/openssl/tls_udp.c:63]: inheritance?

**Comment by alfredh** [comment]
this patch is quite large, around 1100 lines

is it possible to split it up into smaller parts, for easier review ?

**Comment by fAuernigg** [comment]
> this patch is quite large, around 1100 lines
> 
> is it possible to split it up into smaller parts, for easier review ?

I added PR #711, #712, #713.

The libre test added in PR 713 requires new api's for the http server to test renegotiation.
The test requires #712 to fix renegotiation in the tls statemachine under TLS v1.2
The test requires #711 to fix reading certificate files during this test on Win32.

So what I wanted to say is PR #713 will never have the green light for test ok until TLS #712 / #711 was merged.
This PR #713 is to test why PR #712 fixes the tls client and server state machine during tls renegotiation.

**Comment by cspiel1** [comment]
Is replaced by #711, #712, #713.
